sys/net/gnrc/link_layer/lorawan/gnrc_lorawan.c

```diff
@@ -0,0 +1,349 @@
+            mac->state = LORAWAN_STATE_RX_2;
+            break;
+        case LORAWAN_STATE_RX_2:
+            gnrc_lorawan_mlme_no_rx(mac);
+            gnrc_lorawan_mcps_event(mac, MCPS_EVENT_NO_RX, 0);
+            mac->state = LORAWAN_STATE_IDLE;
+            gnrc_lorawan_mac_release(mac);
+            break;
+        default:
+            assert(false);
+    }
+    _sleep_radio(mac);
+}
+
+/* This function uses a precomputed table to calculate time on air without
+ * using floating point arithmetics */
+static uint32_t lora_time_on_air(size_t payload_size, uint8_t dr, uint8_t cr)
+{
+    assert(dr <= LORAMAC_DR_6);
+    uint8_t _K[6][4] = {    { 0, 1, 5, 5 },
+                            { 0, 1, 4, 5 },
+                            { 1, 5, 5, 5 },
+                            { 1, 4, 5, 4 },
+                            { 1, 3, 4, 4 },
+                            { 1, 2, 4, 3 } };
+
+    uint32_t t_sym = 1 << (15 - dr);
+    uint32_t t_preamble = (t_sym << 3) + (t_sym << 2) + (t_sym >> 2);
+
+    int index = (dr < LORAMAC_DR_6) ? dr : LORAMAC_DR_5;
+    uint8_t n0 = _K[index][0];
+    int nb_symbols;
+
+    uint8_t offset = _K[index][1];
+    if (payload_size < offset) {
+        nb_symbols = 8 + n0 * cr;
+    }
+    else {
+        uint8_t c1 = _K[index][2];
+        uint8_t c2 = _K[index][3];
+        uint8_t pos = (payload_size - offset) % (c1 + c2);
+        uint8_t cycle = (payload_size - offset) / (c1 + c2);
+        nb_symbols = 8 + (n0 + 2 * cycle + 1 + (pos > (c1 - 1))) * cr;
+    }
+
+    uint32_t t_payload = t_sym * nb_symbols;
+    return t_preamble + t_payload;
+}
+
+void gnrc_lorawan_send_pkt(gnrc_lorawan_t *mac, gnrc_pktsnip_t *pkt, uint8_t dr)
+{
+    mac->state = LORAWAN_STATE_TX;
+
+    iolist_t iolist = {
+        .iol_base = pkt->data,
+        .iol_len = pkt->size,
+        .iol_next = (iolist_t *) pkt->next
+    };
+
+    uint32_t chan = gnrc_lorawan_pick_channel(mac);
+    _config_radio(mac, chan, dr, false);
+
+    mac->last_dr = dr;
+
+    uint8_t cr;
+    netdev_get_pass(&mac->netdev, NETOPT_CODING_RATE, &cr, sizeof(cr));
+
+    mac->toa = lora_time_on_air(gnrc_pkt_len(pkt), dr, cr + 4);
+
+    if (netdev_send_pass(&mac->netdev, &iolist) == -ENOTSUP) {
+        DEBUG("gnrc_lorawan: Cannot send: radio is still transmitting");
+    }
+
+}
+
+void gnrc_lorawan_process_pkt(gnrc_lorawan_t *mac, gnrc_pktsnip_t *pkt)
+{
+    mac->state = LORAWAN_STATE_IDLE;
+    xtimer_remove(&mac->rx);
+
+    uint8_t *p = pkt->data;
+
+    uint8_t mtype = (*p & MTYPE_MASK) >> 5;
+    switch (mtype) {
+        case MTYPE_JOIN_ACCEPT:
+            gnrc_lorawan_mlme_process_join(mac, pkt);
+            break;
+        case MTYPE_CNF_DOWNLINK:
+        case MTYPE_UNCNF_DOWNLINK:
+            gnrc_lorawan_mcps_process_downlink(mac, pkt);
+            break;
+        default:
+            gnrc_pktbuf_release(pkt);
+            break;
```

Comment:
you don't use break on `default` case in other places, be consistent.

Reply:
I added all missing `break` statements in the last push